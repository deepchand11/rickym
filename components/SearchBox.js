import React, { useEffect, useState, useCallback } from "react";
import { StyleSheet, View, TextInput } from "react-native";
import { AntDesign } from "@expo/vector-icons";
import { TouchableOpacity } from "react-native-gesture-handler";

const SearchBox = ({ setInfo, setCharacters, setLoading }) => {
  const [text, setText] = useState("");

  const handleSearchPress = async () => {
    if (text.length > 0) {
      try {
        setLoading(true);
        const response = await fetch(
          "https://rickandmortyapi.com/api/character/?name=" + text
        );
        const data = await response.json();
        if (!data.error && data.results.length > 0) {
          setLoading(false);
          setInfo(data.info);
          setCharacters([...data.results]);
        } else {
          setLoading(false);
          setCharacters([]);
        }
      } catch (err) {
        setLoading(false);
        console.log(err);
      }
    }
  };

  return (
    <View style={styles.header}>
      <TextInput
        style={styles.input}
        onChangeText={setText}
        value={text}
        testID="searchId"
      />
      <TouchableOpacity onPress={handleSearchPress} testID="searchBtnId">
        <View style={styles.btnSearch}>
          <AntDesign name="search1" size={24} color="white" />
        </View>
      </TouchableOpacity>
    </View>
  );
};

export default SearchBox;

const styles = StyleSheet.create({
  input: {
    height: 40,
    margin: 12,
    backgroundColor: "#ffffff",
    borderRadius: 10,
    flex: 1,
    padding: 8,
  },
  btnSearch: {
    backgroundColor: "#FF813F",
    borderRadius: 3,
    height: 40,
    width: 40,
    alignItems: "center",
    justifyContent: "center",
  },
  header: {
    padding: 5,
    flexDirection: "row",
    alignItems: "center",
  },
});
