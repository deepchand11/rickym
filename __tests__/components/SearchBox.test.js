import "react-native";
import React from "react";

// Note: test renderer must be required after react-native.
import renderer from "react-test-renderer";
import SearchBox from "../../components/SearchBox";
import { fireEvent, waitFor, render } from "@testing-library/react-native";

beforeEach(() => {
  fetch.mockClear();
});

it("Search trigger success", async () => {
  const { getByTestId } = render(
    <SearchBox
      setInfo={() => null}
      setCharacters={() => null}
      setLoading={() => null}
    />
  );
  const handleSearchPress = fetch.mockResponseSuccess(
    JSON.stringify({ data: { info: {} }, results: [{}] })
  );

  const button = getByTestId("searchBtnId");
  fireEvent.press(button);
  //   await waitFor(() => expect(handleSearchPress).toHaveBeenCalledTimes(0));
});

it("Search trigger error", async () => {
  const { getByTestId, toJSON } = render(
    <SearchBox
      setInfo={() => null}
      setCharacters={() => null}
      setLoading={() => null}
    />
  );
  toJSON();
  const searchTxt = "Rick";
  const handleSearchPress = fetch.mockResponseFailure(
    JSON.stringify({ data: { error: {} } })
  );
  const search = getByTestId("searchId");
  fireEvent.changeText(search, searchTxt);
  const button = getByTestId("searchBtnId");
  fireEvent.press(button);
});

it("Search trigger success with no data", async () => {
  const { getByTestId } = render(
    <SearchBox
      setInfo={() => null}
      setCharacters={() => null}
      setLoading={() => null}
    />
  );
  const searchTxt = "Rick";
  const handleSearchPress = fetch.mockResponseSuccess(
    JSON.stringify({ data: { results: [] } })
  );
  const search = getByTestId("searchId");
  fireEvent.changeText(search, searchTxt);
  const button = getByTestId("searchBtnId");
  fireEvent.press(button);
});

it("Search component renders correctly", () => {
  const tree = renderer
    .create(
      <SearchBox
        setInfo={() => null}
        setCharacters={() => null}
        setLoading={() => null}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
