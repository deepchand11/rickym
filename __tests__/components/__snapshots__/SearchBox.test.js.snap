// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search component renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flexDirection": "row",
      "padding": 5,
    }
  }
>
  <TextInput
    allowFontScaling={true}
    onChangeText={[Function]}
    rejectResponderTermination={true}
    style={
      Object {
        "backgroundColor": "#ffffff",
        "borderRadius": 10,
        "flex": 1,
        "height": 40,
        "margin": 12,
        "padding": 8,
      }
    }
    testID="searchId"
    underlineColorAndroid="transparent"
    value=""
  />
  <RNGestureHandlerButton
    collapsable={false}
    onGestureEvent={[Function]}
    onGestureHandlerEvent={[Function]}
    onGestureHandlerStateChange={[Function]}
    onHandlerStateChange={[Function]}
    rippleColor={0}
    testID="searchBtnId"
  >
    <View
      accessible={true}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#FF813F",
            "borderRadius": 3,
            "height": 40,
            "justifyContent": "center",
            "width": 40,
          }
        }
      >
        <Text
          allowFontScaling={false}
          style={
            Array [
              Object {
                "color": "white",
                "fontSize": 24,
              },
              undefined,
              Object {
                "fontFamily": "anticon",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          î™°
        </Text>
      </View>
    </View>
  </RNGestureHandlerButton>
</View>
`;
